docs/architecture/adr-054-semver-compatible-modules.md

```diff
@@ -0,0 +1,493 @@
+# ADR 054: Semver Compatible SDK Modules
+
+## Changelog
+
+* 2022-04-27: First draft
+
+## Status
+
+PROPOSED
+
+## Abstract
+
+In order to move the Cosmos SDK to a system of decoupled semantically versioned
+modules which can be composed in different combinations (ex. staking v3 with
+bank v1 and distribution v2), we need to reassess how we organize the API surface
+of modules to avoid problems with go semantic import versioning and
+circular dependencies. This ADR explores various approaches we can take to
+addressing these issues.
+
+## Context
+
+There has been [a fair amount of desire](https://github.com/cosmos/cosmos-sdk/discussions/10162)
+in the community for semantic versioning in the SDK and there has been significant
+movement to splitting SDK modules into [standalone go modules](https://github.com/cosmos/cosmos-sdk/issues/11899).
+Both of these will ideally allow the ecosystem to move faster because we won't
+be waiting for all dependencies to update synchronously. For instance, we could
+have 3 versions of the core SDK compatible with the latest 2 releases of
+CosmWasm as well as 4 different versions of staking . This sort of setup would
+allow early adopters to aggressively integrate new versions, while allowing 
+more conservative users to be selective about which versions they're ready for.
+
+In order to achieve this, we need to solve the following problems:
+1. because of the way [go semantic import versioning](https://research.swtch.com/vgo-import) (SIV)
+works, moving to SIV naively will actually make it harder to achieve these goals
+2. circular dependencies between modules need to be broken to actually release
+many modules in the SDK independently
+3. pernicious minor version incompatibilities introduced through correctly
+[evolving protobuf schemas](https://developers.google.com/protocol-buffers/docs/proto3#updating)
+without correct [unknown field filtering](./adr-020-protobuf-transaction-encoding.md#unknown-field-filtering)
+
+Note that all the following discussion assumes that the proto file versioning and state machine versioning of a module
+are distinct in that:
+* proto files are maintained in a non-breaking way (using something like [buf breaking](https://docs.buf.build/breaking/overview)
+  to ensure all changes are backwards compatible)
+* proto file versions get bumped much less frequently, i.e. we might maintain `cosmos.bank.v1` through many versions
+  of the bank module state machine
+* state machine breaking changes are more common and ideally this is what we'd want to semantically version with
+go modules, ex. `x/bank/v2`, `x/bank/v3`, etc.
+
+### Problem 1: Semantic Import Versioning Compatibility
+
+Consider we have a module `foo` which defines the following `MsgDoSomething` and that we've released its state
+machine in go module `example.com/foo`:
```

Comment:
A protobuf package named `foo.v1` would require a Go module named `example.com/foo/v1`.

Reply:
That is not how our code base is setup. I think it should be but it isn't. And a protobuf package doesn't necessitate its own go module. They can be bundled with other code